src/python/pants/rules/core/fmt.py

```diff
@@ -44,20 +47,19 @@ class Fmt(Goal):
   )
 
   for result in results:
-    files_content = await Get(FilesContent, Digest, result.digest)
-    # TODO: This is hacky and inefficient, and should be replaced by using the Workspace type
-    # once that is available on master.
-    # Blocked on: https://github.com/pantsbuild/pants/pull/8329
-    for file_content in files_content:
-      with Path(get_buildroot(), file_content.path).open('wb') as f:
-        f.write(file_content.content)
-
+    # NB: we cannot call `workspace.materialize_directories()` in one single call with all of the
```

Comment:
Rather than looping like this, you could merge the `Snapshots` with `DirectoriesToMerge` (which will succeed if all colliding files end up with the same output) and then materialize them. If they did not have the same output, it would be because of an interesting case like "T1 owns file F and has settings S1, while T2 also owns file F and has settings S2, and they are incompatible", and that would be a nice thing to catch before beginning to mutate the workspace.

Reply:
Good idea! How will this work with multiple autoformatters, though? If Black changes `test.py` and isort also changes `test.py`, I believe the current implementation will safely handle that. 

If we use `DirectoriesToMerge`, would that fail when both digests change `test.py`?

Reply:
> Good idea! How will this work with multiple autoformatters, though? If Black changes test.py and isort also changes test.py, I believe the current implementation will safely handle that.

I don't know how it would, unfortunately. If they run concurrently, one of them will lose, and the result will not be stable. You'd have to feed the output of one into the other.

Reply:
Okay. I will make a ticket for the remaining design work around this and then update this to use merged directories. Thank you!

Reply:
Cool. I sanity checked that this new implementation still supports multiple targets owning the same source file, which implies that we would have multiple digests with the same change. `DirectoriesToMerge` correctly handles this case. I'm assuming the engine will only fail when there are conflicting changes that cannot be resolved. Same principle as a git merge conflict.